cmd/openfga/openfga_test.go

```diff
@@ -0,0 +1,919 @@
+//go:build functional
+// +build functional
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/openfga/openfga/pkg/id"
+	"github.com/openfga/openfga/pkg/testutils"
+	"github.com/stretchr/testify/require"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"google.golang.org/grpc"
+	grpcbackoff "google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	healthv1pb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+type OpenFGATester interface {
+	GetGRPCPort() string
+	GetHTTPPort() string
+	Cleanup() func()
+}
+
+type serverHandle struct {
+	grpcPort string
+	httpPort string
+	cleanup  func()
+}
+
+func (s *serverHandle) GetGRPCPort() string {
+	return s.grpcPort
+}
+
+func (s *serverHandle) GetHTTPPort() string {
+	return s.httpPort
+}
+
+func (s *serverHandle) Cleanup() func() {
+	return s.cleanup
+}
+
+// newOpenFGATester spins up an openfga container with the default service ports
+// exposed for testing purposes. Before running functional tests it is assumed
+// the openfga/openfga container is already built and available to the docker engine.
+func newOpenFGATester(t *testing.T, args ...string) (OpenFGATester, error) {
+	t.Helper()
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv)
+	require.NoError(t, err)
+
+	cmd := []string{"run"}
+	cmd = append(cmd, args...)
+
+	containerCfg := container.Config{
+		Env: []string{},
+		ExposedPorts: nat.PortSet{
+			nat.Port("8080/tcp"): {},
+			nat.Port("8081/tcp"): {},
+			nat.Port("3000/tcp"): {},
+		},
+		Image: "openfga/openfga",
+		Cmd:   cmd,
+	}
+
+	hostCfg := container.HostConfig{
+		AutoRemove:      true,
+		PublishAllPorts: true,
+		PortBindings: nat.PortMap{
+			"8080/tcp": []nat.PortBinding{},
+			"8081/tcp": []nat.PortBinding{},
+			"3000/tcp": []nat.PortBinding{},
+		},
+	}
+
+	ulid, err := id.NewString()
+	require.NoError(t, err)
+
+	name := fmt.Sprintf("openfga-%s", ulid)
+
+	cont, err := dockerClient.ContainerCreate(context.Background(), &containerCfg, &hostCfg, nil, nil, name)
```

Comment:
Should we create a single ctx instead of using `context.Background()` everywhere?

Reply:
Sure!